packages/modules/providers/locking-redis/src/services/redis-lock.ts

```diff
@@ -0,0 +1,269 @@
+    this.redisClient.defineCommand("acquireLock", {
+      numberOfKeys: 1,
+      lua: `
+        local key = KEYS[1]
+        local ownerId = ARGV[1]
+        local ttl = tonumber(ARGV[2])
+        local awaitQueue = ARGV[3] == 'true'
+
+        local setResult = redis.call('SET', key, ownerId, 'EX', ttl, 'NX')
+
+        if setResult then
+          return 1
+        elseif not awaitQueue then
+          -- Key already exists; retrieve the current ownerId
+          local currentOwnerId = redis.call('GET', key)
+          if currentOwnerId == ownerId then
+            redis.call('SET', key, ownerId, 'EX', ttl, 'XX')
+            return 1
+          else
+            return 0
+          end
+        else
+          return 0
+        end
+      `,
+    })
+
+    // Define the custom command for releasing locks
+    this.redisClient.defineCommand("releaseLock", {
+      numberOfKeys: 1,
+      lua: `
+        local key = KEYS[1]
+        local ownerId = ARGV[1]
+
+        if redis.call('GET', key) == ownerId then
+          return redis.call('DEL', key)
+        else
+          return 0
+        end
+      `,
+    })
+  }
+
+  private getKeyName(key: string): string {
+    return `${this.keyNamePrefix}${key}`
+  }
+
+  async execute<T>(
+    keys: string | string[],
+    job: () => Promise<T>,
+    args?: {
+      timeout?: number
+    }
+  ): Promise<T> {
+    const timeout = Math.max(args?.timeout ?? this.waitLockingTimeout, 1)
+    const timeoutSeconds = Number.isNaN(timeout) ? 1 : timeout
+
+    const cancellationToken = { cancelled: false }
+    const promises: Promise<any>[] = []
+    if (timeoutSeconds > 0) {
+      promises.push(this.getTimeout(timeoutSeconds, cancellationToken))
+    }
+
+    promises.push(
+      this.acquire_(
+        keys,
+        {
+          awaitQueue: true,
+        },
+        cancellationToken
+      )
+    )
+
+    await Promise.race(promises)
+
+    try {
+      return await job()
+    } finally {
+      await this.release(keys)
+    }
+  }
+
+  async acquire(
+    keys: string | string[],
+    args?: {
+      ownerId?: string
+      expire?: number
+      awaitQueue?: boolean
+    }
+  ): Promise<void> {
+    return this.acquire_(keys, args)
+  }
+
+  async acquire_(
+    keys: string | string[],
+    args?: {
+      ownerId?: string
+      expire?: number
+      awaitQueue?: boolean
+    },
+    cancellationToken?: { cancelled: boolean }
+  ): Promise<void> {
+    keys = Array.isArray(keys) ? keys : [keys]
+
+    const timeout = Math.max(args?.expire ?? this.waitLockingTimeout, 1)
+    const timeoutSeconds = Number.isNaN(timeout) ? 1 : timeout
+    let retryTimes = 0
+
+    const ownerId = args?.ownerId ?? "*"
+    const awaitQueue = args?.awaitQueue ?? false
+
+    const acquirePromises = keys.map(async (key) => {
+      const errMessage = `Failed to acquire lock for key "${key}"`
+      const keyName = this.getKeyName(key)
+
+      const acquireLock = async () => {
+        while (true) {
```

Comment:
thought: For good measure, should we have a max retries config?

Reply:
This number of retries is tricky because it will greatly depend on the time to wait for the locking and the interval of the retry.
We are increasing the time for the next retry to not retry constantly if there are many process trying to to lock it.
As we always have a timeout set, and this scenario is an edge case I believe this is fine.

Reply:
Got it, makes sense.